Backport of:

From c54baf1009cb6cc0de8c39aa09b5b5b27f294854 Mon Sep 17 00:00:00 2001
From: Cristy <urban-warrior@imagemagick.org>
Date: Sun, 5 Jun 2016 14:19:46 -0400
Subject: [PATCH] RLE check for pixel offset less than 0

Heap overflow report from Craig Young

(cherry picked from commit 73fb0aac5b958521e1511e179ecc0ad49f70ebaf)

origin: upstream, https://github.com/ImageMagick/ImageMagick/commit/73fb0aac5b958521e1511e179ecc0ad49f70ebaf
bug-debian: https://bugs.debian.org/833744
---
 coders/rle.c | 10 ++++++----
 1 file changed, 6 insertions(+), 4 deletions(-)

Index: imagemagick-6.7.7.10/coders/rle.c
===================================================================
--- imagemagick-6.7.7.10.orig/coders/rle.c	2016-11-14 11:11:47.937342423 -0500
+++ imagemagick-6.7.7.10/coders/rle.c	2016-11-14 11:13:47.133347981 -0500
@@ -182,11 +182,11 @@
     number_planes,
     number_planes_filled,
     one,
-    offset,
     pixel_info_length;
 
   ssize_t
     count,
+    offset,
     y;
 
   unsigned char
@@ -391,7 +391,8 @@
           offset=((image->rows-y-1)*image->columns*number_planes)+x*
             number_planes+plane;
           operand++;
-          if (offset+((size_t) operand*number_planes) > pixel_info_length)
+          if ((offset < 0) ||
+              (offset+((size_t) operand*number_planes) > pixel_info_length))
             {
               if (number_colormaps != 0)
                 colormap=(unsigned char *) RelinquishMagickMemory(colormap);
@@ -423,13 +424,14 @@
             x*number_planes+plane;
           offset=((image->rows-y-1)*image->columns*number_planes)+x*
             number_planes+plane;
-          p=rle_pixels+offset;
-          if (offset+((size_t) operand*number_planes) > pixel_info_length)
+          if ((offset < 0) ||
+              (offset+((size_t) operand*number_planes) > pixel_info_length))
             {
               if (number_colormaps != 0)
                 colormap=(unsigned char *) RelinquishMagickMemory(colormap);
               ThrowReaderException(CorruptImageError,"UnableToReadImageData");
             }
+          p=rle_pixels+offset;
           for (i=0; i < (ssize_t) operand; i++)
           {
             if ((y < (ssize_t) image->rows) &&
